Cookbook/Technical-Documents/DeepLinks.md

```diff
@@ -0,0 +1,209 @@
+    let url: URL
+    let options: [UIApplication.OpenURLOptionsKey : Any]
+    let intent: ApplicationInvocation.Intent?
+```
+
+Incoming deep link events are then processed in the private `handle` method, producing a routing event, bound to the `routes` variable in the invocation router protocol:
+
+``` swift
+protocol ApplicationInvocationRouterProtocol {
+    var routes: Signal<RoutingEvent, NoError> { get }
+}
+```
+
+The produced routes are then plugged into `BabylonTabBarViewModel` (always present when the user is logged in) where we define which ones require a clinical access check first, and which can be handled right away. If the route requires clinical access, the user is presented a password prompt. If, for some reason, the password check fails, the route is discarded, thus maintaining secure access to the patient's data.
+
+`BabylonTabBarViewModel` in its turn produces a `Route.external(RoutingEvent, .deeplink)` event handled through its Flow controller, always presented modally over the tab bar controller. Some deep links switch tabs instead of presenting screens modally, which is encapsulated in the `RoutingEvent` as well.
+
+Overall, the initial `DeepLink` value gets transformed into a `RoutingEvent` which is, essentially, just like a regular `Route` handled by Flow controllers in other parts of the app, and in `BabylonTabBarFlowController` these events are also handled in a regular fashion by constructing a screen with a `Builder` and presenting it.
+
+### Prevent certain deep links from being handled
+
+As a measure to release support for new deep links when the feature is fully complete, which is not necessarily at the time when the deep link is introduced, we have a special protocol:
+
+``` swift
+protocol DeeplinkConfiguration {
+    func isEnabled(for intent: ApplicationInvocation.Intent) -> Bool
+}
+```
+
+Any app target can define a simple `var deeplinkConfiguration: DeeplinkConfiguration { get }` as part of its `AppConfiguration` and specify which deep links are not supported by that particular app flavour. These will be filtered out during handling in `ApplicationInvocationRouter` and will have no effect.
+
+The default value is just to return `true`, this enabling all available deep links.
+
+### Open a deep link from within the app
+
+When the deep links found extended usage within the app, it became apparent that interacting with them in the app through `UIApplication.openURL(_:)` was extremely inconvenient during testing. It was common for both testers and developers to have multiple versions of the app installed at the same time, and iOS, as it turned out, randomly chooses the app to handle a particular custom URL scheme from the ones installed, as long as they support it. We were unable to control which app would be opened, when interacting with a deep link from inside an app, and this could even affect end users, since they could, in theory, have several of our apps installed at the same time.
```

Comment:
```suggestion
When we started to use deep links in the app more, it became apparent that interacting with them in the app through `UIApplication.openURL(_:)` was extremely inconvenient during testing. It was common for both testers and developers to have multiple versions of the app installed, and iOS, as it turned out, randomly chooses the app to handle a particular custom URL scheme from the ones installed (as long as they support it). When interacting with a deep link from inside an app we were unable to control which app would be opened, and this could even affect end users, since in theory they could have several of our apps installed.
```